Tests/DependencyInjection/Compiler/MiddlewarePassTest.php

```diff
@@ -0,0 +1,292 @@
+
+        $func($container);
+
+        $container->compile();
+
+        return $container;
+    }
+
+    private function assertMiddlewareInjected(
+        ContainerBuilder $container,
+        string $connName,
+        string $middlewareClass,
+        bool $connectionNameAware = false
+    ): void {
+        $middlewareFound = $this->getMiddlewaresForConn($container, $connName, $middlewareClass);
+
+        $this->assertCount(1, $middlewareFound, sprintf(
+            'Middleware %s not injected in doctrine.dbal.%s_connection.configuration',
+            $middlewareClass,
+            $connName
+        ));
+
+        $callsFound = [];
+        foreach ($middlewareFound[0]->getMethodCalls() as $call) {
+            if ($call[0] !== 'setConnectionName') {
+                continue;
+            }
+
+            $callsFound[] = $call;
+        }
+
+        if ($connectionNameAware) {
+            $this->assertCount(1, $callsFound);
+            $this->assertSame($callsFound[0][1][0] ?? null, $connName);
+        } else {
+            $this->assertCount(0, $callsFound);
+        }
+    }
+
+    private function assertMiddlewareNotInjected(
+        ContainerBuilder $container,
+        string $connName,
+        string $middlewareClass
+    ): void {
+        $middlewareFound = $this->getMiddlewaresForConn($container, $connName, $middlewareClass);
+
+        $this->assertCount(0, $middlewareFound, sprintf(
+            'Middleware %s injected in doctrine.dbal.%s_connection.configuration',
+            $middlewareClass,
+            $connName
+        ));
+    }
+
+    /** @return Definition[] */
+    private function getMiddlewaresForConn(ContainerBuilder $container, string $connName, string $middlewareClass): array
+    {
+        $calls            = $container->getDefinition('conf_' . $connName)->getMethodCalls();
+        $middlewaresFound = [];
+        foreach ($calls as $call) {
+            if ($call[0] !== 'setMiddlewares' || ! isset($call[1][0])) {
+                continue;
+            }
+
+            foreach ($call[1][0] as $middlewareDef) {
+                if ($middlewareDef->getClass() !== $middlewareClass) {
+                    continue;
+                }
+
+                $middlewaresFound[] = $middlewareDef;
+            }
+        }
+
+        return $middlewaresFound;
+    }
+}
+
+class Middleware1
+{
+}
+
+class Middleware2 implements ConnectionNameAwareInterface
+{
+    public function setConnectionName(string $name): void
+    {
+    }
+}
+
+/** @psalm-suppress UndefinedClass */
+if (interface_exists(Middleware::class)) {
+    class Middleware3 implements Middleware
+    {
+        public function wrap(Driver $driver): Driver
+        {
+            return $driver;
+        }
+    }
+
+    if (PHP_VERSION_ID >= 80000) {
+        #[AsMiddleware]
+        class Middleware4
+        {
+        }
+
+        #[AsMiddleware(connections: ['conn2'])]
+        class Middleware5
```

Comment:
I took the freedom to rename it as `AutoconfiguredMiddlewareWithConnection` instead to avoid confusion with `ConnectionNameAwareInterface` (a middleware can target a specific connection without being aware of this connection name); is it ok for you ?